File: 0206-reverse-linked-list/0206-reverse-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        // Initialize two pointers
        ListNode* prev = NULL;
        ListNode* curr = head;

        // Iterate as long as the current pointer is not null
        while (curr) {
            // Store the next node temporarily
            ListNode* temp = curr->next;

            // Reverse the direction of the link for the current node
            curr->next = prev;

            // Shift the pointers forward
            prev = curr;
            curr = temp;
        }

        // After the loop, 'prev' will be pointing to the new head of the reversed list
        return prev;
    }
};

File: 0206-reverse-linked-list/0206-reverse-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        
        ## initialize two pointers
        
        prev = None
        curr = head
        
        ## iterate if the head is not null
        
        while curr:
            temp = curr.next
            ## now this indicates the last element in the list 
            curr.next = prev
            ## shifitng the pointers forward
            prev = curr
            curr = temp
            
        return prev    